### ISSUE TITLE
SelectCheckboxMenu: SelectItem with null value checked by default

### ISSUE THREAD

**Issue by brivalin** [primefaces/primefaces#4712]
## 1) Environment
- PrimeFaces version: 6.2.0
- Does it work on the newest released PrimeFaces version? Version? Not working in 7.0
- Does it work on the newest sources in GitHub? Not working
- Application server + version: Wildfly 10.1.0.Final
- Affected browsers: All

## 2) Expected behavior

Items with null values don't have to be checked by default.

## 3) Actual behavior

Items with null values are checked.

## 4) Steps to reproduce

No Steps to reproduce, only create SelectCheckboxMenu with code samples and you will get this : 

![selectCheckboxMenu](https://user-images.githubusercontent.com/26894018/55326099-42d3b680-5487-11e9-894d-4a11746277da.PNG)

## 5) Sample XHTML
```xhtml
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:ui="http://java.sun.com/jsf/facelets"
      xmlns:f="http://java.sun.com/jsf/core"
      xmlns:h="http://java.sun.com/jsf/html"
      xmlns:p="http://primefaces.org/ui">
	<h:head></h:head> 
	<body>
		<h:form>
			<p:selectCheckboxMenu value="#{testPrimeBean.values}" label="Test" updateLabel="true">
				<f:selectItems value="#{testPrimeBean.items}" />
			</p:selectCheckboxMenu>
		</h:form>
	</body> 
</html>
```
## 6) Sample bean
```java
@Named
@ViewScoped
public class TestPrimeBean implements Serializable{
	private static final long serialVersionUID = 1L;
	
	private List<SelectItem> items;
	private String[] values;

	@PostConstruct
	public void init() {
		items = new ArrayList<>();
		values = new String[0];
		
		items.add(new SelectItem(null, "One"));
		items.add(new SelectItem("two","Two"));
		items.add(new SelectItem("three"));
		items.add(new SelectItem(null));
	}

	public List<SelectItem> getItems() {
		return items;
	}

	public void setItems(List<SelectItem> items) {
		this.items = items;
	}

	public String[] getValues() {
		return values;
	}

	public void setValues(String[] values) {
		this.values = values;
	}
}
```
..

**Comment by melloware** [comment]
I think this may be done on purpose and is generic for all "select" type components.  This same code is called by all the Select components and i believe it says if your value is null and the selectItem option is NULL then the answer is "true" its selected.

https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/renderkit/SelectRenderer.java#L244-L246

So its selecting the items you have marked as NULL because your backing bean value is NULL thus its considering them a match.

**Comment by brivalin** [comment]
Thanks for your quick reply @melloware.

I have made some tests on SelectManyMenu and SelectCheckboxMenu, and you are right, they have the same behavior with my sample.
But I also tried to use a list instead of an array by replacing String[] by List<String> and I have seen that the behavior is different. In case of list, checkboxes are not selected if selectItem value is null.

I think the best way should be to have the same behavior which is for me to not check in the two cases.

**Comment by melloware** [comment]
Agreed the behavior should be the same for both.

**Comment by melloware** [comment]
@brivalin I can't reproduce.  Attached is my test case:
[pf-4712.zip](https://github.com/primefaces/primefaces/files/3031198/pf-4712.zip)

Can you run "mvn clean jetty:run-exploded" and navigate to http://localhost:8080/primefaces-test/test.xhtml.

I am seeing String[] and List<String> show the same behavior.

**Comment by brivalin** [comment]
Perhaps I misunderstood something but if you init the list in your example with :
```java
stringValues = new ArrayList<>();
```
You will not have the same behavior.

**Comment by melloware** [comment]
If you initialize the array you get the same result.

```java
	private String[] values = new String[]{"","","",""};
	private List<String> stringValues = new ArrayList<>();
```

The reason you are seeing that behavior is because in the current example it initializes the String array to `new String[0]`.

I am not sure why but it looks like it has been like this for years in the SelectManyRenderer.

https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/renderkit/SelectManyRenderer.java#L60-L67

Basically if its a java.util.Collection it is returned however if the Array [] has no values it returns NULL.  To be consistent the Collection would have to be checked for .isEmpty() and return NULL to match the behavior of the Array.

I would like @mertsincan @tandraschko @Rapster to weigh in on this.

**Comment by tandraschko** [comment]
no idea
can you compare with JSF impls or check the history?

**Comment by melloware** [comment]
According to the history it has been that way since OptimusPrime created the file on Dec 1, 2011.

https://github.com/primefaces/primefaces/commit/8ffa0c9eeb1a577d0e30917db2676266d07c0006#diff-7c884a51c67b9f8062f0f67ff02b0684

**Comment by tandraschko** [comment]
IMO if the "value" attribute, doesnt matter if array or collection, is empty - nothing should be selected per default

**Comment by melloware** [comment]
In that case then I believe the problem is these lines.

https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/renderkit/SelectRenderer.java#L227-L229

If says if the value is NULL and the SelectItem is Null it marks it selected.

**Comment by tandraschko** [comment]
yep, i >feel< that its wrong for selectMany

**Comment by melloware** [comment]
Agreed I will submit a PR.

**Comment by tandraschko** [comment]
could you please check my commit @melloware?
I fixes this problem and is a small perf improvement as it avoids new arrays.

**Comment by melloware** [comment]
Looks good to me and appears to fix the issue.
